cardano-crypto-class/src/Cardano/Crypto/DirectSerialise.hs

```diff
@@ -0,0 +1,179 @@
+-- swap, which is an important requirement for KES.
+module Cardano.Crypto.DirectSerialise
+where
+
+import Foreign.Ptr
+import Foreign.C.Types
+import Control.Monad (when)
+import Control.Monad.Class.MonadThrow (MonadThrow)
+import Control.Monad.Class.MonadST (MonadST, stToIO)
+import Data.STRef (newSTRef, readSTRef, writeSTRef)
+import Cardano.Crypto.Libsodium.Memory (copyMem)
+
+-- | Direct deserialization from raw memory.
+--
+-- @directDeserialise f@ should allocate a new value of type 'a', and
+-- call @f@ with a pointer to the raw memory to be filled. @f@ may be called
+-- multiple times, for data structures that store their data in multiple
+-- non-contiguous blocks of memory.
+--
+-- The order in which memory blocks are visited matters.
+class DirectDeserialise a where
+  directDeserialise :: (MonadST m, MonadThrow m) => (Ptr CChar -> CSize -> m ()) -> m a
+
+-- | Direct serialization to raw memory.
+--
+-- @directSerialise f x@ should call @f@ to expose the raw memory underyling
+-- @x@. For data types that store their data in multiple non-contiguous blocks
+-- of memory, @f@ may be called multiple times, once for each block.
+--
+-- The order in which memory blocks are visited matters.
+class DirectSerialise a where
+  directSerialise :: (MonadST m, MonadThrow m) => (Ptr CChar -> CSize -> m ()) -> a -> m ()
+
+-- | Helper function for bounds-checked serialization.
+-- Verifies that no more than the maximum number of bytes are written, and
+-- returns the actual number of bytes written.
+directSerialiseTo :: forall m a.
+                     DirectSerialise a
+                  => MonadST m
+                  => MonadThrow m
+                  => (Int -> Ptr CChar -> CSize -> m ())
+                  -> Int
+                  -> a
+                  -> m Int
+directSerialiseTo writeBytes dstsize val = do
+  posRef <- stToIO $ newSTRef 0
+  let pusher :: Ptr CChar -> CSize -> m ()
+      pusher src srcsize = do
+        pos <- stToIO $ readSTRef posRef
+        let pos' = pos + fromIntegral srcsize
+        when (pos' > dstsize) (error "Buffer overrun")
+        writeBytes pos src (fromIntegral srcsize)
+        stToIO $ writeSTRef posRef pos'
+  directSerialise pusher val
+  stToIO $ readSTRef posRef
+
+-- | Helper function for size-checked serialization.
+-- Verifies that exactly the specified number of bytes are written.
+directSerialiseToChecked :: forall m a.
+                            DirectSerialise a
+                         => MonadST m
+                         => MonadThrow m
+                         => (Int -> Ptr CChar -> CSize -> m ())
+                         -> Int
+                         -> a
+                         -> m ()
+directSerialiseToChecked writeBytes dstsize val = do
+  bytesWritten <- directSerialiseTo writeBytes dstsize val
+  when (bytesWritten /= dstsize) $
+    error $ "directSerialiseToChecked: size check failed, expected " ++ show dstsize ++ ", but found " ++ show bytesWritten
```

Comment:
I was curious in another place as well (in particular: `(error "Buffer overrun")`). Why are we using `error` for deserialization?

I'd expect deserialization to be applied to untrusted input, so this computation could result in an exception that is not suppose to be handled gracefully.

Is it because these functions are designed for testing and another mechanism will be used in production? If these are functions suitable for production maybe we should create a custom exception and utilize `MonadThrow` instead?

Reply:
Good point.

My thinking was that when the size check fails, it's going to be either a malicious input, or a bug; either way, we cannot continue gracefully, and the reasonable thing to do is to fail loudly.

We can still do that with a custom exception though, so I'll do that instead.